File: BFS---Breadth-First-Search.py
def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start)  # Processing node

    for neighbor in graph[start]:
        if neighbor not in visited:
            dfs(graph, neighbor, visited)

# Example usage
graph = {
    0: [1, 2],
    1: [2],
    2: [0, 3],
    3: [3]
}
dfs(graph, 2)

File: BFS---Breadth-First-Search.js
function bfs(graph, startNode) {
    const visited = new Set();
    const queue = [startNode];

    while (queue.length > 0) {
        const node = queue.shift(); // Remove and get the first element in the queue

        if (!visited.has(node)) {
            visited.add(node);
            console.log(node); // Processing node

            const neighbors = graph[node];
            for (const neighbor of neighbors) {
                if (!visited.has(neighbor)) {
                    queue.push(neighbor);
                }
            }
        }
    }
}

// Example usage
const graph = {
    0: [1, 2],
    1: [2],
    2: [0, 3],
    3: [3]
};
bfs(graph, 2);
